tiledb/api/src/group/mod.rs

```diff
@@ -0,0 +1,780 @@
+use std::ops::Deref;
+
+use crate::config::{Config, RawConfig};
+use crate::context::{ContextBound, ObjectType};
+use crate::error::Error;
+use crate::key::LookupKey;
+use crate::{Context, Datatype};
+
+extern crate tiledb_sys as ffi;
+use crate::string::{RawTDBString, TDBString};
+use crate::Result as TileDBResult;
+pub type QueryType = crate::array::Mode;
+use crate::metadata::Metadata;
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct GroupInfo {
+    pub uri: String,
+    pub group_type: ObjectType,
+    pub name: String,
+}
+
+pub(crate) struct RawGroup {
+    ffi: *mut ffi::tiledb_group_t,
+}
+
+impl RawGroup {
+    pub fn new(ffi: *mut ffi::tiledb_group_t) -> Self {
+        RawGroup { ffi }
+    }
+}
+
+impl Deref for RawGroup {
+    type Target = *mut ffi::tiledb_group_t;
+    fn deref(&self) -> &Self::Target {
+        &self.ffi
+    }
+}
+
+impl Drop for RawGroup {
+    fn drop(&mut self) {
+        unsafe { ffi::tiledb_group_free(&mut self.ffi) }
+    }
+}
+
+#[derive(ContextBound)]
+pub struct Group<'ctx> {
+    #[context]
+    context: &'ctx Context,
+    raw: RawGroup,
+}
+
+impl<'ctx> Group<'ctx> {
+    pub(crate) fn capi(&self) -> *mut ffi::tiledb_group_t {
+        *self.raw
+    }
+
+    pub(crate) fn new(context: &'ctx Context, raw: RawGroup) -> Self {
+        Group { context, raw }
+    }
+
+    pub fn create<S>(context: &'ctx Context, name: S) -> TileDBResult<()>
+    where
+        S: AsRef<str>,
+    {
+        let c_name = cstring!(name.as_ref());
+        context.capi_return(unsafe {
+            ffi::tiledb_group_create(context.capi(), c_name.as_ptr())
+        })
+    }
+
+    pub fn open<S>(
+        context: &'ctx Context,
+        uri: S,
+        query_type: QueryType,
+        config: Option<&Config>,
+    ) -> TileDBResult<Self>
+    where
+        S: AsRef<str>,
+    {
+        let c_context = context.capi();
+        let mut group_raw: *mut ffi::tiledb_group_t = out_ptr!();
+
+        let c_uri = cstring!(uri.as_ref());
+
+        context.capi_return(unsafe {
+            ffi::tiledb_group_alloc(c_context, c_uri.as_ptr(), &mut group_raw)
+        })?;
+
+        if let Some(cfg) = config {
+            let c_cfg = cfg.capi();
+            context.capi_return(unsafe {
+                ffi::tiledb_group_set_config(c_context, group_raw, c_cfg)
+            })?;
+        }
+
+        let raw_group = RawGroup::new(group_raw);
+        let query_type_raw = query_type.capi_enum();
+        context.capi_return(unsafe {
+            ffi::tiledb_group_open(c_context, group_raw, query_type_raw)
+        })?;
+
+        let mut c_open: i32 = out_ptr!();
+        context
+            .capi_return(unsafe {
+                ffi::tiledb_group_is_open(c_context, raw_group.ffi, &mut c_open)
+            })
+            .expect("TileDB internal error when checking for open group.");
```

Comment:
`expect` isn't appropriate here since we can return `Result`

also I think you can remove this if you want, your comment yesterday (earlier this week? something) in `Drop` convinced me that you did it correctly the first time